### PR TITLE
Add (Prox-)SVRG as a solver for GLMs

### PR DESCRIPTION
**Pull request by bagibence**
Add a preliminary implementation of [SVRG](https://papers.nips.cc/paper_files/paper/2013/hash/ac1dd209cbcc5e5d1c6e28598e8cbbe8-Abstract.html) and [Prox-SVRG](https://arxiv.org/abs/1403.4699v1) compatible with the jaxopt API, so nemos can use it as a solver:

- the `run` method (used in `GLM.fit`) assumes the full data is loaded into memory and runs optimization until convergence.
- the `update` method (used in `GLM.update`) updates the parameters on a single data point or batch. Note that currently, when fitting on batches and calling `glm.update`, some steps of the algorithm have to be implemented outside.

`ProxSVRG` takes a `prox` argument for the proximal operator which should be a callable with the same signature as the functions defined in `jaxopt.prox` (e.g. `jaxopt.prox.prox_lasso`).
`SVRG` inherits from `ProxSVRG`, uses the identity as the proximal operator and adjusts some calls to account for the different number of arguments.

### CODE REVIEW

File: src/nemos/solvers.py [hunk 1]
```diff
@@ -0,0 +1,726 @@
...
+            ind = random.randint(subkey, (self.batch_size,), 0, N)
+
+            # perform a single update on the mini-batch or data point
+            xk = self._xk_update(
+                xk,
+                xs,
+                df_xs,
+                state.stepsize,
+                prox_lambda,
+                *(tree_slice(arg, ind) for arg in args),
```
Comment (BalzaniEdoardo) [inline src/nemos/solvers.py:517]:
```suggestion
                *tree_slice(args, ind),
```

This is sufficient because you are sure that args is a tuple, the output of a tree_map is going to be a tuple (same tree struct as args), and you can unpack output of tree slice directly.

File: src/nemos/solvers.py [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+        state = state._replace(
+            iter_num=state.iter_num + 1,
+            key=key,
+            x_av=tree_scalar_mul(1 / m, x_sum),
+        )
+
+        # the next anchor point is the parameters at the end of the inner loop
+        # or the average over the inner loop
+        return OptStep(params=xk, state=state)
+        # return OptStep(params=state.x_av, state=state)
```
Comment (BalzaniEdoardo) [inline src/nemos/solvers.py:547]:
```suggestion
```

File: src/nemos/solvers.py [hunk 3]
```diff
@@ -0,0 +1,726 @@
...
+
+    # @staticmethod
+    # def _error(x, x_prev, stepsize):
+    #    diff_norm = tree_l2_norm(tree_sub(x, x_prev))
+    #    return diff_norm / stepsize
```
Comment (BalzaniEdoardo) [inline src/nemos/solvers.py:552]:
```suggestion
```

File: src/nemos/solvers.py [hunk 4]
```diff
@@ -0,0 +1,726 @@
...
+    @staticmethod
+    def _error(x, x_prev, stepsize):
+        return tree_l2_norm(tree_sub(x, x_prev)) / tree_l2_norm(x_prev)
+
+    # @staticmethod
+    # def _error(x, x_prev, stepsize):
+    #    # adapted from scikit-learn's SAG solver
+    #    diff = tree_sub(x, x_prev)
+    #    flat_diff, _ = jax.flatten_util.ravel_pytree(diff)
+    #    return jnp.max(jnp.abs(flat_diff))
```
Comment (BalzaniEdoardo) [inline src/nemos/solvers.py:563]:
```suggestion
```

File: src/nemos/solvers.py [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+    def loss_fn(params, X, y):
+        ...
+
+    svrg = SVRG(loss_fn)
+    params, state = svrg.run(init_params, X, y)
+
+    References
+    ----------
+    Prox-SVRG - https://arxiv.org/abs/1403.4699v1
+    SVRG - https://papers.nips.cc/paper_files/paper/2013/hash/ac1dd209cbcc5e5d1c6e28598e8cbbe8-Abstract.html
```
Comment (BalzaniEdoardo) [inline src/nemos/solvers.py:599]: reference as in the suggestion of ProxSVRG

File: src/nemos/solvers.py [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+        ----------
+        init_params : Any
+            Initial parameters to start from.
+        args:
+            Positional arguments passed to loss function `fun` and its gradient.
+            For GLMs it is:
+                X :
+                    Input data.
+                y :
+                    Output data.
```
Comment (BalzaniEdoardo) [inline src/nemos/solvers.py:706]: remember to update the docstrings as in proxSVRG

File: src/nemos/solvers_naive.py
Comment (BalzaniEdoardo) [inline src/nemos/solvers_naive.py:1]: should we remove this? I don't think we are using it

**Review by BalzaniEdoardo — CHANGES_REQUESTED** [review]
This is looking great already. My comments are pretty minor, very good job! 

This PR is quite big already, so I would not add a tutorial for the docs in this one, we should open a dedicated PR for it.

I'll add an issue describing what I am envisioning for that!

File: src/nemos/solvers_naive.py
Comment (bagibence) [reply src/nemos/solvers_naive.py:1]: Yes, I'll remove that.

File: src/nemos/solvers.py [hunk 7]
```diff
@@ -0,0 +1,726 @@
...
+
+      objective(params, hyperparams_prox, *args, **kwargs) =
+        fun(params, *args, **kwargs) + non_smooth(params, hyperparams_prox)
+
+    Attributes
+    ----------
+    fun: Callable
+        smooth function of the form ``fun(x, *args, **kwargs)``.
+    prox: Callable
+        proximity operator associated with the function ``non_smooth``.
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py:52]: I think I copied it from jaxopt, but I also like proximal better.

File: src/nemos/solvers.py [hunk 8]
```diff
@@ -0,0 +1,726 @@
...
+KeyArrayLike = ArrayLike
+
+
+class SVRGState(NamedTuple):
+    iter_num: int
+    key: KeyArrayLike
+    error: float
+    stepsize: float
+    # loss_log: jnp.ndarray
+    xs: Optional[tuple] = None
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py:33]: I think it could be any pytree, it was just `(Ws, b)` for GLMs.

File: src/nemos/solvers.py (hunk 1 above)
Comment (bagibence) [reply src/nemos/solvers.py:517]: Cool!

File: src/nemos/solvers.py [hunk 9]
```diff
@@ -0,0 +1,726 @@
...
+    df_xs: Optional[tuple] = None
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py:34]: I also like explicit names, but in this case I might prefer these as they are closer to the math and the pseudo-code in the papers. How about explaining what they are in multiple places?

File: src/nemos/solvers.py [hunk 10]
```diff
@@ -0,0 +1,726 @@
...
+        state : SVRGState
+            Initialized optimizer state
+        """
+        df_xs = None
+        if init_full_gradient:
+            df_xs = self.loss_gradient(init_params, *args)
+
+        state = SVRGState(
+            iter_num=0,
+            key=self.key if self.key is not None else random.key(0),
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py:138]: Done.

File: src/nemos/solvers.py (hunk 3 above)
Comment (bagibence) [reply src/nemos/solvers.py]: Removed.

File: src/nemos/solvers.py (hunk 4 above)
Comment (bagibence) [reply src/nemos/solvers.py]: Removed.

File: src/nemos/solvers.py [hunk 11]
```diff
@@ -0,0 +1,726 @@
...
+                    Output data.
+
+
+        Returns
+        -------
+        OptStep
+            xs : Any
+                Parameters after taking one step defined in the inner loop of Prox-SVRG.
+            state : SVRGState
+                Updated state.
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py:233]: Added.

File: src/nemos/solvers.py [hunk 12]
```diff
@@ -0,0 +1,726 @@
...
+        hyperparams_prox : float
+            Parameters of the proximal operator, in our case the regularization strength.
+            Not used here, but required to be consistent with the jaxopt API.
+        args:
+            Positional arguments passed to loss function `fun` and its gradient.
+            For GLMs it is:
+                X :
+                    Input data.
+                y :
+                    Output data.
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py]: I changed this everywhere. Let me know if it's good like this. I also thought about having the full description only in one place, then referring to it from other methods. How would you like that?

File: src/nemos/solvers.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py]: Changed it to `Pytree`.

File: src/nemos/solvers.py [hunk 13]
```diff
@@ -0,0 +1,726 @@
...
+
+        Parameters
+        ----------
+        loss_log : jnp.ndarray
+            1D array storing the loss values.
+        i : int
+            Index at which to update, most likely the current iteration number.
+        params : Any
+            Parameters with which to evaluate the loss.
+        args:
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py]: Done.

File: src/nemos/solvers.py [hunk 14]
```diff
@@ -0,0 +1,726 @@
...
+        OptStep
+            xs : Any
+                Parameters at the end of the last inner loop.
+                (... or the average of the parameters over the last inner loop)
+            state : SVRGState
+                Updated state.
+        """
+        n_points_per_arg = {leaf.shape[0] for leaf in jax.tree.leaves(args)}
+        if not len(n_points_per_arg) == 1:
+            raise ValueError("All arguments must have the same sized first dimension.")
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py:475]: I just tried to measure this by moving it up into the `_run` method, and I don't think it matters, so I would leave it here.
I removed the calculations for `x_av` which is never used now, and that might actually save some time, but I didn't measure that.

File: src/nemos/solvers.py [hunk 15]
```diff
@@ -0,0 +1,726 @@
...
+            For GLMs it is:
+                X :
+                    Input datapoint or mini-batch.
+                y :
+                    Output datapoint or mini-batch.
+        """
+        dfik_xk = self.loss_gradient(xk, *args)
+        dfik_xs = self.loss_gradient(xs, *args)
+
+        gk = jax.tree_util.tree_map(lambda a, b, c: a - b + c, dfik_xk, dfik_xs, df_xs)
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py]: Added the comment.

File: src/nemos/solvers.py [hunk 16]
```diff
@@ -0,0 +1,726 @@
...
+        args:
+            Hyperparameters passed to `prox` and positional arguments passed to loss function `fun` and its gradient.
+            For GLMs it is:
+                X :
+                    Input datapoint or mini-batch.
+                y :
+                    Output datapoint or mini-batch.
+        """
+        dfik_xk = self.loss_gradient(xk, *args)
+        dfik_xs = self.loss_gradient(xs, *args)
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py]: I renamed these to `grad_of_fik_at_xk` and `grad_of_fik_at_xs`. We might want to rename `df_xs` to `grad_of_f_at_xs` as well. These might be a good compromise that's kind of close to that math but easier to interpret. What do you think?

I also renamed `final_xs` to `final_params` where it occurred.

File: src/nemos/solvers.py [hunk 17]
```diff
@@ -0,0 +1,726 @@
...
+                    Output data.
+
+        Returns
+        -------
+        Updated loss log.
+        """
+        return loss_log.at[i].set(self.fun(params, *args))
+
+    @partial(jit, static_argnums=(0,))
+    def _xk_update(
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py]: I renamed this to `_xk_update_step` to signal that it's a single gradient step. I think I'm leaning towards leaving it like this instead of saying parameter because it's more explicit which step of the SVRG algorithm it refers to, whereas parameter step/update is a bit more vague to me. I feel like it might also discourage misuse or confusion with `update`.

File: src/nemos/solvers.py [hunk 18]
```diff
@@ -0,0 +1,726 @@
...
+                    Input datapoint or mini-batch.
+                y :
+                    Output datapoint or mini-batch.
+        """
+        dfik_xk = self.loss_gradient(xk, *args)
+        dfik_xs = self.loss_gradient(xs, *args)
+
+        gk = jax.tree_util.tree_map(lambda a, b, c: a - b + c, dfik_xk, dfik_xs, df_xs)
+
+        next_xk = tree_add_scalar_mul(xk, -stepsize, gk)
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py:190]: How about `x_k_plus_one` / `x__k_plus_one`?

File: src/nemos/solvers.py [hunk 19]
```diff
@@ -0,0 +1,725 @@
...
+from jax import grad, jit, lax, random
+from jaxopt import OptStep
+from jaxopt._src import loop
+from jaxopt.prox import prox_none
+
+from .tree_utils import tree_add_scalar_mul, tree_l2_norm, tree_slice, tree_sub
+from .typing import KeyArrayLike, Pytree
+
+
+class SVRGState(NamedTuple):
```
Comment (billbrod) [inline src/nemos/solvers.py:16]: should we be using a NamedTuple for this or a [dataclass](https://docs.python.org/3/library/dataclasses.html)?

From [this stackoverflow](https://stackoverflow.com/questions/51671699/data-classes-vs-typing-namedtuple-primary-use-cases), it looks like NamedTuples make sense when the data structure is immutable, but the state does get changed, right?

File: src/nemos/solvers.py [hunk 20]
```diff
@@ -0,0 +1,725 @@
...
+        Number of data points to sample per inner loop iteration.
+
+    Example
+    -------
+    def loss_fn(params, X, y):
+        ...
+
+    svrg = ProxSVRG(loss_fn, prox_fun)
+    params, state = svrg.run(init_params, prox_lambda, X, y)
+
```
Comment (billbrod) [inline src/nemos/solvers.py:61]: Example is not rendered correctly. I believe it should be [formatted like](https://numpydoc.readthedocs.io/en/latest/format.html#examples):
```suggestion
    Examples
    --------
    >>> def loss_fn(params, X, y):
    >>>    ...
    >>>
    >>> svrg = ProxSVRG(loss_fn, prox_fun)
    >>> params, state = svrg.run(init_params, prox_lambda, X, y)

```

File: src/nemos/solvers.py [hunk 21]
```diff
@@ -0,0 +1,725 @@
...
+        init_full_gradient: bool = False,
+    ) -> SVRGState:
+        """
+        Initialize the solver state
+
+        Parameters
+        ----------
+        init_params :
+            Pytree containing the initial parameters.
+            For GLMs it's a tuple of (W, b)
```
Comment (billbrod) [inline src/nemos/solvers.py:101]: this docstirng is irrelevant, right? shouldn't it be describing what the SVRG needs?

File: src/nemos/solvers.py [hunk 22]
```diff
@@ -0,0 +1,725 @@
...
+        hyperparams_prox :
```
Comment (billbrod) [inline src/nemos/solvers.py:102]: since this is unused, should we raise an error or warning if it gets set?

File: src/nemos/solvers.py [hunk 23]
```diff
@@ -0,0 +1,725 @@
...
+from .typing import KeyArrayLike, Pytree
+
+
+class SVRGState(NamedTuple):
+    iter_num: int
+    key: KeyArrayLike
+    error: float
+    stepsize: float
+    xs: Optional[Pytree] = None
+    df_xs: Optional[Pytree] = None
```
Comment (billbrod) [inline src/nemos/solvers.py:22]: I think this naming is unclear. is it documented anywhere? can we give this class a docstring explaining what they are? and is this naming matching a convention? if so, should point that out. if not, I think it should be changed to something more clear

File: src/nemos/solvers.py [hunk 24]
```diff
@@ -0,0 +1,725 @@
...
+        prox_lambda: Union[float, None],
+        *args,
+    ) -> OptStep:
+        """
+        Perform a single parameter update on the passed data (no random sampling or loops)
+        and increment `state.iter_num`.
+
+        Please note that this is called by `GLM.update`, but repeated calls to `GLM.update`
+        on mini-batches passed to it will not result in running the full (Prox-)SVRG,
+        and parts of the algorithm will have to be implemented outside.
```
Comment (billbrod) [inline src/nemos/solvers.py:211]: - currently, this gets called by `GLM.update`, but it will get called by any model update step, right? I would say something like "This gets called by `BaseRegressor.solver_update` (e.g., as called by `GLM.update`)..."
- For "repeated calls ... will not result in running the full SVRG": in order to run the full SVRG, they should call `.run`, right? We should say that instead of making it sound like they need to implement it themselves

File: src/nemos/solvers.py [hunk 25]
```diff
@@ -0,0 +1,725 @@
...
+
+        Returns
+        -------
+        OptStep
+            xs :
+                Parameters after taking one step defined in the inner loop of Prox-SVRG.
+            state :
+                Updated state.
+        """
+        # NOTE this doesn't update state.xs, state.df_xs, that has to be done outside
```
Comment (billbrod) [inline src/nemos/solvers.py:296]: this is already in the docstring, so can be removed

File: src/nemos/solvers.py [hunk 26]
```diff
@@ -0,0 +1,725 @@
...
+        self,
+        current_params: Pytree,
+        state: SVRGState,
+        prox_lambda: Union[float, None],
+        *args,
+    ) -> OptStep:
+        """
+        Update parameters given a mini-batch of data and increment iteration/epoch number in state.
+
+        Note that this method doesn't update state.xs, state.df_xs, that has to be done outside.
```
Comment (billbrod) [inline src/nemos/solvers.py:266]: would clarify: "e.g., as is done in ..."

File: src/nemos/solvers.py [hunk 27]
```diff
@@ -0,0 +1,725 @@
...
+    @partial(jit, static_argnums=(0,))
+    def run(
+        self,
+        init_params: Pytree,
+        prox_lambda: Union[float, None],
+        *args,
+    ) -> OptStep:
+        """
+        Run a whole optimization until convergence or until `maxiter` epochs are reached.
+        Called from `GLM.fit` and assumes that X and y are the full data set.
```
Comment (billbrod) [inline src/nemos/solvers.py:317]: same point about GLM being too specific

File: src/nemos/solvers.py [hunk 28]
```diff
@@ -0,0 +1,725 @@
...
+    def _run(
+        self,
+        init_params: Pytree,
+        init_state: SVRGState,
+        prox_lambda: Union[float, None],
+        *args,
+    ) -> OptStep:
+        """
+        Run a whole optimization until convergence or until `maxiter` epochs are reached.
+        Called from `GLM.fit` and assumes that X and y are the full data set.
```
Comment (billbrod) [inline src/nemos/solvers.py:365]: same point about GLM being too specific

File: src/nemos/solvers.py [hunk 29]
```diff
@@ -0,0 +1,725 @@
...
+    stepsize : float
+        Constant step size to use.
+    tol: float
+        Tolerance level for the error when comparing parameters
+        at the end of consecutive epochs to check for convergence.
+    batch_size: int
+        Number of data points to sample per inner loop iteration.
+
+    Example
+    -------
```
Comment (billbrod) [inline src/nemos/solvers.py:571]: same point about this not rendering correctly

File: src/nemos/tree_utils.py [hunk 30]
```diff
@@ -140,3 +141,65 @@ def pytree_map_and_reduce(
     cond_tree = jax.tree_util.tree_map(map_fn, *pytrees, is_leaf=is_leaf)
     # for some reason, tree_reduce doesn't work well with any.
     return reduce_fn(jax.tree_util.tree_leaves(cond_tree))
+
+
+def tree_slice(data: Any, idx):
```
Comment (billbrod) [inline src/nemos/tree_utils.py:146]: I thought slicing already worked on our pytrees, why was this necessary?

**Review by billbrod — CHANGES_REQUESTED** [review]
I don't understand the algorithm, so I'm not looking at its details here. I'm trusting you and Edoardo have that taken care of!

Most of my comments are about documentation.

- Without the example, it's a little hard for me to see how this all works, but I agree with Edoardo that it's fine to have a separate PR for that. I don't see an issue for that right now -- @bagibence can you create one and drop any relevant notes in there before you leave?
- Why are the tree changes in this PR as well? Are they related to the solver?

File: src/nemos/solvers.py (hunk 19 above)
Comment (bagibence) [reply src/nemos/solvers.py:16]: The state does get updated, but by creating a new one (`._replace` creates a copy) instead of modifying the original one.
I basically just copied [how jaxopt defines optimizer states](https://github.com/google/jaxopt/blob/6d882923a54feb6f68b46193d9787cc73bfdadd3/jaxopt/_src/proximal_gradient.py#L84), which might have been inspired by [this example](https://jax.readthedocs.io/en/latest/stateful-computations.html#simple-worked-example-linear-regression).

File: src/nemos/solvers.py (hunk 20 above)
Comment (bagibence) [reply src/nemos/solvers.py]: Changed it.

File: src/nemos/solvers.py (hunk 29 above)
Comment (bagibence) [reply src/nemos/solvers.py]: Updated.

File: src/nemos/solvers.py (hunk 22 above)
Comment (bagibence) [reply src/nemos/solvers.py:102]: What do you mean by it getting set?

[This is how jaxopt treats this](https://github.com/google/jaxopt/blob/6d882923a54feb6f68b46193d9787cc73bfdadd3/jaxopt/_src/proximal_gradient.py#L179). I wanted to be consistent with them.

File: src/nemos/solvers.py (hunk 24 above)
Comment (bagibence) [reply src/nemos/solvers.py:211]: - Re `BaseRegressor._solver_update`: Yes, I'll change it.
- Re `.run`: `.run` and `.fit` assumes you have the full data in memory and can index into it. If you don't, you have to rely on `.update` and you do have to implement some parts of the SVRG algorithm yourself. This is in contrast with SGD (shown in the batching example), where you just have to call `.update` on each new minibatch. We have tests that the full SVRG algorithm calling `.update`, but we still need to make a tutorial in the docs. In the future this would be hidden from the user with an alternative interface for batched fitting. These might be relevant:
https://github.com/flatironinstitute/nemos/issues/194
https://github.com/flatironinstitute/nemos/issues/203

Comment (bagibence) [reply src/nemos/solvers.py]: I addressed the first point. Let me know how you like the new description.

File: src/nemos/solvers.py (hunk 27 above)
Comment (bagibence) [reply src/nemos/solvers.py]: Changed it. Let me know if you like it the new way.

File: src/nemos/solvers.py (hunk 28 above)
Comment (bagibence) [reply src/nemos/solvers.py]: Changed it. Let me know if you like it the new way.

File: src/nemos/solvers.py (hunk 25 above)
Comment (bagibence) [reply src/nemos/solvers.py]: Removed it.

File: src/nemos/solvers.py (hunk 26 above)
Comment (bagibence) [reply src/nemos/solvers.py:266]: Do you mean "as is done in `ProxSVRG.run`"? Or how `state.xs` and `state.df_xs` can be updated?

File: src/nemos/solvers.py (hunk 23 above)
Comment (bagibence) [reply src/nemos/solvers.py:22]: They are documented in the implementation of the algorithm, but we can add a docstring here as well.
The naming is trying to match the pseudocode provided in [this paper](https://arxiv.org/abs/2010.00892).

File: src/nemos/tree_utils.py (hunk 30 above)
Comment (bagibence) [reply src/nemos/tree_utils.py:146]: It's for slicing any pytree, not just `FeaturePytree`.

File: src/nemos/solvers.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py]: I renamed it to `_inner_loop_param_update_step`

File: src/nemos/solvers.py [hunk 31]
```diff
@@ -0,0 +1,726 @@
...
+        Updated loss log.
+        """
+        return loss_log.at[i].set(self.fun(params, *args))
+
+    @partial(jit, static_argnums=(0,))
+    def _xk_update(
+        self,
+        xk: Any,
+        xs: Any,
+        df_xs: Any,
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py]: Changed it.

File: src/nemos/solvers.py [hunk 32]
```diff
@@ -0,0 +1,726 @@
...
+        Returns
+        -------
+        Updated loss log.
+        """
+        return loss_log.at[i].set(self.fun(params, *args))
+
+    @partial(jit, static_argnums=(0,))
+    def _xk_update(
+        self,
+        xk: Any,
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py]: Renamed to `params` everywhere.

File: src/nemos/solvers.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py]: I ended up with `next_params`.

File: src/nemos/solvers.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py]: We ended up calling them `minibatch_grad_at_current_params` and `minibatch_grad_at_reference_point` with some comments about how they relate to the pseudocode in Gower et al. 2020.

File: src/nemos/solvers.py [hunk 33]
```diff
@@ -0,0 +1,726 @@
...
+        on mini-batches passed to it will not result in running the full (Prox-)SVRG,
+        and parts of the algorithm will have to be implemented outside.
+
+        Parameters
+        ----------
+        current_params : Any
+            Parameters at the end of the previous update, used as the starting point for the current update.
+        state : SVRGState
+            Optimizer state at the end of the previous update.
+            Needs to have the current anchor point (xs) and the gradient at the anchor point (df_xs) already set.
```
[the remark this answers is not in the document]
Comment (bagibence) [reply src/nemos/solvers.py]: Done.

File: src/nemos/solvers.py (hunk 6 above)
Comment (bagibence) [reply src/nemos/solvers.py]: Updated!

File: src/nemos/solvers.py (hunk 22 above)
Comment (billbrod) [reply src/nemos/solvers.py:136]: I think that's fine, but what I meant was that if these are specified, we should raise a warning or error. as in `if hyperparams_prox is not None`. but actually, I think I'm confused -- why isn't this argument optional?

File: src/nemos/solvers.py (hunk 24 above)
Comment (billbrod) [reply src/nemos/solvers.py]: ah that's right, I forgot about that. as long as we remember to show it in the tutorial, that's fine.

File: src/nemos/solvers.py (hunk 26 above)
Comment (billbrod) [reply src/nemos/solvers.py]: I think this is related to the tutorial for `update` that we're going to add. If I understand correctly, the updating "that has to be done outside" isn't currently shown anywhere (except for a test), right? so this will just need to point to that new tutorial that we write

File: src/nemos/solvers.py [hunk 34]
```diff
@@ -14,12 +14,39 @@
...
+    key :
+        Random key to use when sampling data points or mini-batches.
+    error :
+        Scaled difference (~distance) between subsequent parameter values
+        used to monitor convergence.
+    stepsize :
+        Step size of the individual gradient steps.
+    reference_point :
+        Anchor/reference/snapshot point where the full gradient is calculated in the SVRG algorithm.
+        Corresponds to `x_{s}` in the pseudocode in [1]
```
[the remark this answers is not in the document]
Comment (billbrod) [reply src/nemos/solvers.py:33]: @BalzaniEdoardo did you figure this out for mkdocs?

**Review by billbrod — CHANGES_REQUESTED** [review]
This generally looks very good! Mainly some minor changes here, and I responded in a couple of my previous comments, so make sure to check those again.

Also, and I know this is annoying, but `test_solvers` still has references to `xs` and `df_xs`. I think that should be updated to match the new nomenclature, which will make it easier for us in the future.

File: src/nemos/solvers.py (hunk 22 above)
Comment (BalzaniEdoardo) [reply src/nemos/solvers.py:136]: unsure why, but in the general `jaxopt.ProximalGradient` API, both `init_state` and `update` require a `hyperparameter_prox` input. 

This is an annoying choice, they should have make it keyword and optional, but in their case it made sense to have an hyperparameter variable at state initialization since you cannot anticipate if a specific proximal method would require it or not.

For us, I don't think we should care about this parameter for this ProxSVRG since we are not implementing a base class for any stochastic solver (which we should, but not in this PR). I'll remove it.

An idea for a refinement of this solver class is to have a general interface for stochastic solvers (with inner/outer loop structure ) and the inner loop should be an abstract class, as well as equipping the class with a `run_iterator` for batching. I'll add this idea to Bence's issue #194.

File: src/nemos/solvers.py (hunk 26 above)
Comment (BalzaniEdoardo) [reply src/nemos/solvers.py]: yes, I agree

Comment (BalzaniEdoardo) [reply src/nemos/solvers.py]: I added to the issue on the SVRG tutorial

File: src/nemos/solvers.py (hunk 34 above)
[the remark this answers is not in the document]
Comment (BalzaniEdoardo) [reply src/nemos/solvers.py]: I figured that our links for refs did not work because the References is not considered a markdown section. I did make it a section and this seems to be rendered well. Do you like this solution?

File: src/nemos/solvers.py [hunk 35]
```diff
@@ -30,15 +30,15 @@ class SVRGState(NamedTuple):
         Step size of the individual gradient steps.
     reference_point :
         Anchor/reference/snapshot point where the full gradient is calculated in the SVRG algorithm.
-        Corresponds to `x_{s}` in the pseudocode in [1]
+        Corresponds to `x_{s}` in the pseudocode[$^{[1]}$](#references).
```
Comment (billbrod) [inline src/nemos/solvers.py:33]:
```suggestion
        Corresponds to $x_{s}$ in the pseudocode[$^{[1]}$](#references).
```

**Review by billbrod — APPROVED** [review]
Two small issues and then this is ready to merge!

- Docstring of `ProxSVRG` mentions `hyperparams_prox`, which is no longer an argument. So that should be removed.

File: src/nemos/solvers.py (hunk 34 above)
[the remark this answers is not in the document]
Comment (billbrod) [reply src/nemos/solvers.py]: I think this solution is fine for now, but we should figure out how to get them rendering correctly. Not as part of this PR, I opened an issue for it: #212

**Comment by BalzaniEdoardo** [comment]
> * `hyperparams_prox`

The docstrings are ok, because that is referring to any hyperparameter that a general proximal operator gets as input. For example, what we call regularization_strength is one such hyper-parameter for ridge.

**Review by BalzaniEdoardo — APPROVED** [review]
